master/internal/task/allocation.go

```diff
@@ -171,224 +206,353 @@ func NewAllocation(
-		ctx.Tell(a.req.AllocationRef, msg)
-		a.Error(ctx, msg)
+		})
+	}
+
+	return a.rendezvous.watch(rID)
+}
 
+// UnwatchRendezvous removes a rendezvous watcher.
+func (a *allocation) UnwatchRendezvous(rID sproto.ResourcesID) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.rendezvous.unwatch(rID)
+}
+
+func (a *allocation) RendezvousTimeout() {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	if err := a.rendezvous.checkTimeout(); err != nil {
+		a.sendTaskLog(&model.TaskLog{Log: err.Error()})
+	}
+}
+
+func (a *allocation) validateRendezvous() error {
+	if a.rendezvous != nil {
+		return nil
+	}
+
+	if len(a.resources) == 0 {
+		return ErrAllocationUnfulfilled{Action: "rendezvous"}
+	}
+
+	switch a.resources.first().Summary().ResourcesType {
+	case sproto.ResourcesTypeDockerContainer, sproto.ResourcesTypeK8sPod:
+		break
 	default:
-		a.Error(ctx, actor.ErrUnexpectedMessage(ctx))
+		return ErrBehaviorUnsupported{Behavior: "rendezvous"}
 	}
+
 	return nil
 }
 
-// RequestResources sets up the allocation.
-func (a *Allocation) RequestResources(ctx *actor.Context) error {
+// awaitTermination waits for the allocation and any goroutines associated with to exit.
+func (a *allocation) awaitTermination() *AllocationExited {
+	a.wg.Wait()
+	return a.exited
+}
+
+// requestResources sets up the allocation.
+func (a *allocation) requestResources() (*sproto.ResourcesSubscription, error) {
 	if a.req.Restore {
 		// Load allocation.
-		ctx.Log().Debug("RequestResources load allocation")
+		a.syslog.Debug("requestResources load allocation")
 		err := db.Bun().NewSelect().Model(&a.model).
 			Where("allocation_id = ?", a.model.AllocationID).
 			Scan(context.TODO())
 		if err != nil {
-			return errors.Wrap(err, "loading trial allocation")
+			return nil, errors.Wrap(err, "loading trial allocation")
 		}
 	} else {
 		// Insert new allocation.
-		ctx.Log().Debug("RequestResources add allocation")
+		a.syslog.Debug("requestResources add allocation")
 
 		a.setModelState(model.AllocationStatePending)
 		if err := a.db.AddAllocation(&a.model); err != nil {
-			return errors.Wrap(err, "saving trial allocation")
+			return nil, errors.Wrap(err, "saving trial allocation")
 		}
 	}
 
-	a.req.AllocationRef = ctx.Self()
-	if err := a.rm.Allocate(ctx, a.req); err != nil {
-		return errors.Wrap(err, "failed to request allocation")
+	sub, err := a.rm.Allocate(a.system, a.req)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request allocation")
 	}
 	a.sendTaskLog(&model.TaskLog{
-		Log: fmt.Sprintf("Scheduling %s (id: %s)", a.req.Name, ctx.Self().Parent().Address().Local()),
+		Log: fmt.Sprintf("Scheduling %s (id: %s)", a.req.Name, a.req.AllocationID),
 	})
-	return nil
+	return sub, nil
 }
 
-// Cleanup ensures an allocation is properly closed. It tries to do everything before failing and
+// Close ensures an allocation is properly closed. It tries to do everything before failing and
 // ensures we don't leave any resources running.
```

Comment:
Nit: in plain English, "closing" a door makes sense. Closing things like files or channels makes sense (we're done reading/writing). I don't know what closing an allocation means without reading the method's code. Maybe `CleanUp` was a better name? Or maybe we should beef up the doc comment.

Reply:
renamed back, i agree. wanted to make it a proper close for other reasons but didnt get around to it.